autoscheduler/frontend/src/tests/CourseSelectCard.test.tsx

```diff
@@ -9,98 +14,183 @@ import { Provider } from 'react-redux';
+    test('when we change the selected course', async () => {
+      // arrange
+      fetchMock.mockResponseOnce(JSON.stringify({
+        results: ['CSCE 121', 'CSCE 221', 'CSCE 312'],
+      }));
+      fetchMock.mockImplementationOnce(testFetch);
+      fetchMock.mockResponseOnce(JSON.stringify({
+        results: ['MATH 151'],
+      }));
+      fetchMock.mockImplementationOnce(testFetch);
+
+      const nodeProps = Object.create(Node.prototype, {});
+      // @ts-ignore
+      document.createRange = (): Range => ({
+        setStart: (): void => {},
+        setEnd: (): void => {},
+        commonAncestorContainer: {
+          ...nodeProps,
+          nodeName: 'BODY',
+          ownerDocument: document,
+        },
+      });
+      const store = createStore(autoSchedulerReducer, applyMiddleware(thunk));
+      const {
+        getByText, getByLabelText, getAllByText,
+      } = render(
+        <Provider store={store}><CourseSelectCard id={0} /></Provider>,
+      );
+
+      // act
+      // fill in course
+      const courseEntry = getByLabelText('Course') as HTMLInputElement;
+      act(() => { fireEvent.change(courseEntry, { target: { value: 'CSCE ' } }); });
+      const csce121Btn = await waitForElement(() => getByText('CSCE 121'));
+      act(() => { fireEvent.click(csce121Btn); });
+
+      // switch to section view
+      act(() => { fireEvent.click(getByText('Section')); });
+      const course1Sections = (await waitForElement(() => getAllByText(/50\d/))).length;
+
+      // change course and read sections again
+      act(() => { fireEvent.change(courseEntry, { target: { value: 'MATH 15' } }); });
+      const math151Btn = await waitForElement(() => getByText('MATH 151'));
+      act(() => { fireEvent.click(math151Btn); });
+      const course2Sections = (await waitForElement(() => getAllByText(/51\d/))).length;
+
+      // assert
+      expect(course1Sections).not.toEqual(0);
+      expect(course2Sections).not.toEqual(0);
+      expect(course1Sections).not.toEqual(course2Sections);
+    });
+  });
+
+  describe('only fetches api/sections once', () => {
+    describe('when we search and go to the Sections tab', () => {
+      test('and collapse then expand the card', async () => {
+        // arrange
+        let sectionsFetchCount = 0; // how many times api/sections has been called
+
+        fetchMock.mockImplementation((route: string): Promise<Response> => {
+          if (route.match(/api\/course\/search.+/)) {
+            return Promise.resolve(new Response(JSON.stringify({
+              results: ['CSCE 121'],
+            })));
+          }
+
+          if (route.match(/api\/sections.+/)) {
+            sectionsFetchCount += 1;
+            return testFetch(route);
+          }
+
+          return Promise.resolve(new Response('404 Not Found'));
+        });
+
+        const nodeProps = Object.create(Node.prototype, {});
+        // @ts-ignore
+        document.createRange = (): Range => ({
```

Comment:
Yeah I was thinking the same thing, especially since its not like truly integral the test / just seems to set up the page. What exactly does it do anyways?

Reply:
We have to mock the `createRange` function because the Autocomplete calls it when it makes the popover menu, but the jest "document" doesn't have this function by default